### ISSUE TITLE
Crash on config reload

### ISSUE THREAD

**Issue by travankor** [WayfireWM/wayfire#1120]
**Describe the bug**
Wayfire crashed on config reload

**To Reproduce**
1. Make changes to config
2. Save file
3. ???
4. Wayfire freezes for several seconds before crashing

**Expected behavior**
Don't crash

**Screenshots or stacktrace**
http://ix.io/2SUj

**Wayfire version**
0.7.0

**Comment by droc12345** [comment]
I've had an oddity happen, although wayfire didn't crash, it has to do with editing/saving the config.
I run .git versions of wayfire and wlroots. 

What happened is I had edited the config file (while running wayfire) and when I went to save it, the system appeared to freeze. I had a music player on and it kept playing, but no input keyboard or mouse would do anything.
I ssh'd into it from my laptop and could see that everything was running, but this was being output (repeatedly)

Key repeat discarded, Wayland compositor doesn't seem to be processing events fast enough!

I've not seen this action before and I know that I have edited and saved the config file before this.
While I was ssh'd in, I killed the wayfire process with a SIG HUP and the system came right back (in console mode).
It didn't crash, but I'm wondering if this is somehow related to the original post on crashing.

**Comment by droc12345** [comment]
Previous comments were part right and part wrong. The inotify watch on the .config dir should not exist. And the logic for doing the reload never properly left the while/enclosed for loop and thus stayed hung until wayfire killed.

Proper patch for default-config-backend.cpp

`--- a/src/default-config-backend.cpp    2021-03-29 11:45:59.828084002 -0500
+++ b/src/default-config-backend.cpp    2021-03-29 11:48:26.744993241 -0500
@@ -16,7 +16,6 @@
 
 static void readd_watch(int fd)
 {
-    inotify_add_watch(fd, config_dir.c_str(), IN_CREATE);
     inotify_add_watch(fd, config_file.c_str(), IN_MODIFY);
 }
 
@@ -57,7 +56,7 @@
             {
                 // is file, probably the config file itself
                 should_reload = true;
-                continue;
+                break;
             }
 
             if (config_file == event->name)
@@ -65,6 +64,10 @@
                 should_reload = true;
             }
         }
+        if (should_reload)
+        {
+            break;
+        }
     }
 
     if (should_reload)`

This works, does a reload and comes back to continue running.

**Comment by diamondburned** [comment]
> The inotify watch on the .config dir should not exist.

This was the original behavior; the commit addressed the problem of editing any file in .config would trigger a false reload.

---

`read` blocking is a problem that I probably didn't read the manual enough to realize. I don't think `break` should be used here, though, rather `read` should be made non-blocking.

**Comment by diamondburned** [comment]
Can you try this diff?

```diff
diff --git a/src/default-config-backend.cpp b/src/default-config-backend.cpp
index 7d1651eb..999c44dc 100644
--- a/src/default-config-backend.cpp
+++ b/src/default-config-backend.cpp
@@ -98,7 +98,7 @@ class dynamic_ini_config_t : public wf::config_backend_t
         config = wf::config::build_configuration(
             get_xml_dirs(), SYSCONFDIR "/wayfire/defaults.ini", config_file);
 
-        int inotify_fd = inotify_init1(IN_CLOEXEC);
+        int inotify_fd = inotify_init1(IN_CLOEXEC | IN_NONBLOCK);
         reload_config(inotify_fd);
 
         wl_event_loop_add_fd(wl_display_get_event_loop(display),
```

**Comment by diamondburned** [comment]
I've made pull request #1133 to address the bugs in this issue as a follow-up.

**Comment by droc12345** [comment]
I'll try your patch but I don't think it's going to fix the problem, at least the hanging problem. But it will be tomorrow before I can check.

Watching the directory is ok, but totally unnecessary, all you really care about it the wayfire config file, the .config dir is used by lots of programs and can be written too by any of them, so it's not totally right to check that dir for creating and try and reload because of that.

I added lots of debug statements to see what is going on.
I added the break instead of the continue, because when you get to that point, you have nothing left in the buffer.
You need to break out of the for loop (notice you're still inside the while which isn't correct either.)
So you check should_reload being set to see if the for loop finished and if it's set then do a break (the while loop)

Otherwise you get stuck in a loop with no keyboard or mouse ability to do anything. I had to ssh in to kill the wayfire process to get control back.

**Comment by diamondburned** [comment]
> though from the debug messages, it appears to be reading more often than it should, not sure why

The read loop doesn't stop early to drain the `inotify` buffer.

**Comment by droc12345** [comment]
That wasn't exactly what I meant.

I put some of the debug messages here -> https://dpaste.com/4QMUS6SAB <- (expires 7 days from today)

The first file has 2 parts the 1st is what shows when I edited a file in .config dir that had nothing to do with wayfire.
then the ... separator and the second part is what it shows when I do edit the wayfire config file with vim (which creates a swapfile in the .config dir for the config file being edited) 

Notice in both cases the number of times you see this "Reloading configuration file"

The second file is with the notify on the config dir removed.
It runs "Reloading configuration file" exactly once, which is what you want.

This is still using the patch you have for no blocking.

Just letting you know what I've found.
